File: love2d/src/gameplay/dictionary.lua
local Dictionary = {}
Dictionary.__index = Dictionary

function Dictionary.new(data_dir)
    local self = setmetatable({}, Dictionary)
    self.target_words = {}
    self.allowed_guesses = {}
    self:load_dictionaries(data_dir or "data")
    return self
end

function Dictionary:load_dictionaries(data_dir)
    local target_path = data_dir .. "/target_words.txt"
    local guess_path = data_dir .. "/allowed_guesses.txt"
    
    if love.filesystem.getInfo(target_path) then
        for line in love.filesystem.lines(target_path) do
            local clean = line:match("^%s*(.-)%s*$"):lower()
            if #clean == 5 then
                table.insert(self.target_words, clean)
            end
        end
    end
    
    if love.filesystem.getInfo(guess_path) then
        for line in love.filesystem.lines(guess_path) do
            local clean = line:match("^%s*(.-)%s*$"):lower()
            if #clean == 5 then
                self.allowed_guesses[clean] = true
            end
        end
    end
    
    for _, word in ipairs(self.target_words) do
        self.allowed_guesses[word] = true
    end
    
    if #self.target_words == 0 then
        self.target_words = {"write", "draft", "story", "novel", "print", "press", "paper", "pages"}
        for _, word in ipairs(self.target_words) do
            self.allowed_guesses[word] = true
        end
    end
end

function Dictionary:is_valid_guess(word)
    if not word then return false end
    local clean = word:match("^%s*(.-)%s*$"):lower()
    return self.allowed_guesses[clean] == true
end

function Dictionary:get_random_target()
    local idx = love.math.random(1, #self.target_words)
    return self.target_words[idx]
end

return Dictionary

File: src/gameplay/dictionary.py
import os
import random

class Dictionary:
    def __init__(self, data_dir="data"):
        self.target_words = []
        self.allowed_guesses = set()
        self.load_dictionaries(data_dir)

    def load_dictionaries(self, data_dir):
        target_path = os.path.join(data_dir, "target_words.txt")
        guess_path = os.path.join(data_dir, "allowed_guesses.txt")

        # Load target answers list
        if os.path.exists(target_path):
            with open(target_path, "r", encoding="utf-8") as f:
                self.target_words = [line.strip().lower() for line in f if len(line.strip()) == 5]
        
        # Load allowed guesses list
        if os.path.exists(guess_path):
            with open(guess_path, "r", encoding="utf-8") as f:
                self.allowed_guesses = {line.strip().lower() for line in f if len(line.strip()) == 5}
        
        # Merge target words into allowed guesses to ensure all targets are valid guesses
        self.allowed_guesses.update(self.target_words)

        # Fail-safe defaults if dictionaries are empty
        if not self.target_words:
            self.target_words = ["write", "draft", "story", "novel", "print", "press", "paper", "pages"]
            self.allowed_guesses.update(self.target_words)

    def is_valid_guess(self, word):
        """Checks if a 5-letter word is in the dictionary."""
        return word.strip().lower() in self.allowed_guesses

    def get_random_target(self):
        """Selects a random 5-letter target word."""
        return random.choice(self.target_words)
